src/test/scala/zio/codec/CharCodecSpec.scala

```diff
@@ -0,0 +1,78 @@
+package zio.codec.examples
+
+import zio.Chunk
+import zio.test._
+import zio.test.Assertion._
+import zio.codec.{ CharCodecModule }
+
+object CharCodecSpec extends DefaultRunnableSpec {
+  object CharTest extends CharCodecModule
+  import CharTest._
+
+  def toChunk[A](a: Iterable[A]) = Chunk.fromIterable(a)
+
+  def decodeSuccess[A](c: Codec[A], input: Chunk[Char])(expected: A) =
+    assert(decoder(c)(input))(isRight(equalTo((input.size - 1, expected))))
+
+  val genString1 = Gen.string1(Gen.alphaNumericChar)
+
+  val spec = suite("CharCodecSpec")(
+    testM("token")(
+      check(genString1)(str => decodeSuccess(token(str), toChunk(str))(toChunk(str)))
+    ),
+    testM("token")(
+      check(genString1)(str => decodeSuccess(token(str), toChunk(str))(toChunk(str)))
+    ),
+    testM("tokenAs")(
+      check(Gen.anyInt)(a => decodeSuccess(tokenAs(a.toString, a), toChunk(a.toString))(a))
+    ),
+    testM("~")(
+      check(genString1)(str =>
+        decodeSuccess(token(str) ~ token(str.reverse), toChunk(str + str.reverse))((toChunk(str), toChunk(str.reverse)))
+      )
+    ),
+    testM("|")(
+      check(genString1)(str =>
+        decodeSuccess(token(str) | token(str.reverse), toChunk(str))(Left(toChunk(str))) &&
+          decodeSuccess(token(str) | token("noise" + str.reverse), toChunk("noise" + str.reverse))(
+            Right(toChunk("noise" + str.reverse))
+          )
+      )
+    ),
+    testM("rep")(
+      check(Gen.anyChar, Gen.small(Gen.int(1, _), 1)) { (c, i) =>
+        val chars = toChunk(List.fill(i)(c))
+        decodeSuccess(consume.oneOf(c).rep, chars)(chars)
+      }
+    ),
+    testM("repN")(
+      check(Gen.anyChar, Gen.small(Gen.int(1, _), 1)) { (c, i) =>
+        val chars = toChunk(List.fill(i)(c))
+        decodeSuccess(consume.oneOf(c).repN(i), chars)(chars)
+      }
+    ),
+    testM("opt string")( //this fails index -1 != 0
+      check(genString1) { str =>
+        decodeSuccess((token(str) ~ token("extra").ignore(Chunk.empty)).option, toChunk(str))(
+          Some((toChunk(str), ()))
+        )
```

Comment:
It fails because number of consumed chars is always zero asserted against `str.length`.
The tested codec only succeeds if both `str` **and** `"extra"` are in the input, which is not the case.

Reply:
Thats the strange part. the test below ("opt char") actually changes the index to 0

Reply:
That's right, because "opt string" and "opt char" are very different.

"opt string" asks for _str followed by "extra"_, while "opt char" asks for _str followed by zero or more "e"_.

So "opt string" falls back to `None` at index `-1` because _str + "extra"_ is not found, but "opt char" succeeds to wrap in `Some` at index `0` or more because `str` was actually consumed and exactly zero `"e"` were found.

The same behaviour as "opt string" is seen with "opt char repN" where you ask for _str followed by exact number of "e"_. `"e"` are never found, so fall back index is -1.

The number you see from decoder is "the number of consumer chars".
For "opt string" it will be 0 and for "opt char" it's 1 or more.

Reply:
Oh i see. i assumed `rep` is at least one repetition.